flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java

```diff
@@ -213,6 +218,23 @@ public boolean discard(JobStatus jobStatus) throws Exception {
 
 	void discard() throws Exception {
 		try {
+			// check if we have a shortcut hook to dispose the state
+			final StateDisposeHook disposeHook = this.disposeHook;
+			if (disposeHook != null) {
+				// fast path!
```

Comment:
This method is now a bit more lengthy than it could be; i would suggest to factor out the two paths (fast and standard) in 2 private methods.